src/soc/ibm/power9/homer.c

```diff
@@ -908,22 +1109,350 @@ static void getPpeScanRings(struct xip_hw_header *hw, uint8_t dd)
+		EQ_ANA_BNDY_BUCKET_L3DCC, EQ_ANA_MODE, EQ_ANA_BNDY_BUCKET_26,
+		EQ_ANA_BNDY_BUCKET_27, EQ_ANA_BNDY_BUCKET_28, EQ_ANA_BNDY_BUCKET_29,
+		EQ_ANA_BNDY_BUCKET_30, EQ_ANA_BNDY_BUCKET_31, EQ_ANA_BNDY_BUCKET_32,
+		EQ_ANA_BNDY_BUCKET_33, EQ_ANA_BNDY_BUCKET_34, EQ_ANA_BNDY_BUCKET_35,
+		EQ_ANA_BNDY_BUCKET_36, EQ_ANA_BNDY_BUCKET_37, EQ_ANA_BNDY_BUCKET_38,
+		EQ_ANA_BNDY_BUCKET_39, EQ_ANA_BNDY_BUCKET_40, EQ_ANA_BNDY_BUCKET_41
+	};
+
+	const enum ring_id eq_index_bucket_id = resolve_eq_inex_bucket();
+
+	struct qpmr_header *qpmr_hdr = &homer->qpmr.sgpe.header;
+	struct sgpe_cmn_ring_list *tmp =
+		(void *)&homer->qpmr.sgpe.sram_image[qpmr_hdr->img_len];
+	uint8_t *start = (void *)tmp;
+	uint8_t *payload = tmp->payload;
+
+	uint32_t i = 0;
+
+	for (i = 0; i < sizeof(ring_ids) / sizeof(ring_ids[0]); ++i) {
+		uint8_t this_ring_variant;
+		uint32_t ring_size = MAX_RING_BUF_SIZE;
+
+		enum ring_id id = ring_ids[0];
+		if (id == EQ_INEX)
+			id = eq_index_bucket_id;
+
+		this_ring_variant = ring_variant;
+		if (id == EQ_GPTR         || // EQ GPTR
+		    id == EQ_ANA_GPTR     ||
+		    id == EQ_DPLL_GPTR    ||
+		    id == EX_L3_GPTR      || // EX GPTR
+		    id == EX_L2_GPTR      ||
+		    id == EX_L3_REFR_GPTR ||
+		    id == EQ_TIME         || // EQ TIME
+		    id == EX_L3_TIME      || // EX TIME
+		    id == EX_L2_TIME)
+			this_ring_variant = RV_BASE;
+
+		if ((payload - start) % 8 != 0)
+			payload = start + ALIGN_UP(payload - start, 8);
+
+		if (!tor_access_ring(ring_data->rings_buf, id, PT_SGPE,
+				     this_ring_variant, CACHE0_CHIPLET_ID,
+				     payload, &ring_size, GET_RING_DATA))
+			continue;
+
+		tmp->ring[i] = payload - start;
+		payload += ALIGN_UP(ring_size, 8);
+	}
+
+	qpmr_hdr->common_ring_len = payload - start;
+	qpmr_hdr->common_ring_offset =
+		offsetof(struct homer_st, qpmr.sgpe.sram_image) + qpmr_hdr->img_len;
+}
+
+static void layout_inst_rings_for_sgpe(struct homer_st *homer,
+				       struct ring_data *ring_data,
+				       uint64_t cores, uint32_t ring_variant)
+{
+	struct qpmr_header *qpmr_hdr = &homer->qpmr.sgpe.header;
+	uint32_t inst_rings_offset = qpmr_hdr->img_len + qpmr_hdr->common_ring_len;
+
+	uint8_t *start = &homer->qpmr.sgpe.sram_image[inst_rings_offset];
+	struct sgpe_inst_ring_list *tmp = (void *)start;
+	uint8_t *payload = tmp->payload;
+
+	/* It's EQ_REPR and three pairs of EX rings */
+	const enum ring_id ring_ids[] = {
+		EQ_REPR, EX_L3_REPR, EX_L3_REPR, EX_L2_REPR, EX_L2_REPR,
+		EX_L3_REFR_REPR, EX_L3_REFR_REPR, EX_L3_REFR_TIME,
+		EX_L3_REFR_TIME
+	};
+
+	uint8_t quad = 0;
+
+	for (quad = 0; quad < MAX_QUADS_PER_CHIP; ++quad) {
+		uint8_t i;
+
+		/* Skip non-functional quads */
+		if (!IS_EQ_FUNCTIONAL(quad, cores))
+		    continue;
+
+		for (i = 0; i < sizeof(ring_ids) / sizeof(ring_ids[0]); ++i) {
+			const enum ring_id id = ring_ids[0];
+
+			uint32_t ring_size = MAX_RING_BUF_SIZE;
+
+			uint8_t chiplet_id = CACHE0_CHIPLET_ID + quad*2;
+			if (i != 0 && (i - 1) % 2 == 1)
```

Comment:
Simpler:
```suggestion
			if (i != 0 && (i % 2) == 0)
```
Sidenote: Hostboot has an original idea of odd/even numbers :slightly_smiling_face: